File: node/test/normalize.test.ts
import { readFileSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, it, expect } from 'vitest';
import { normalize } from '../src/quote';
import { normalizeInstrument } from '../src/search';
import type { RawQuote, RawSearchResponse } from '../src/types';

const fixturesDir = resolve(dirname(fileURLToPath(import.meta.url)), '../../fixtures');
const load = (name: string) =>
  JSON.parse(readFileSync(resolve(fixturesDir, name), 'utf8'));

describe('normalize (quote)', () => {
  it('turns the raw quote fixture into the normalized fixture', () => {
    const raw = load('quote.json') as RawQuote;
    const expected = load('quote.normalized.json');
    const q = normalize(raw);

    expect(q).toMatchObject({
      symbol: expected.symbol,
      name: expected.name,
      price: expected.price,
      change: expected.change,
      open: expected.open,
      high: expected.high,
      low: expected.low,
      volume: expected.volume,
      marketCap: expected.marketCap,
      currency: expected.currency,
      exchange: expected.exchange,
    });
    expect(q.changePercent).toBeCloseTo(expected.changePercent, 6);
    expect(q.timestamp).toBeInstanceOf(Date);
    expect(q.timestamp.toISOString()).toBe(expected.timestamp);
  });

  it('returns 0% change when the reference close is 0', () => {
    const raw = load('quote.json') as RawQuote;
    const q = normalize({ ...raw, change: 10, ohlc: { ...raw.ohlc, close: 0 } });
    expect(q.changePercent).toBe(0);
  });
});

describe('normalizeInstrument (search)', () => {
  it('turns the raw search fixture into the normalized results', () => {
    const raw = load('search.json') as RawSearchResponse;
    const expected = load('search.normalized.json');
    expect(raw.instruments.map(normalizeInstrument)).toEqual(expected.results);
  });
});

File: kotlin/bin/test/com/stockseyes/NormalizeTest.kt
package com.stockseyes

import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import java.io.File

/**
 * Normaliser tests — asserts against the shared fixtures.
 *
 * Mirrors `node/test/normalize.test.ts` and `python/tests/test_normalize.py`
 * exactly: load the raw fixture, run the normaliser, and verify the output
 * matches the `*.normalized.json` fixture.
 */
class NormalizeTest {

    private val fixturesDir = File(
        System.getProperty("user.dir")
    ).resolve("../fixtures").canonicalFile

    private fun loadFixture(name: String): String =
        fixturesDir.resolve(name).readText(Charsets.UTF_8)

    private val jsonParser = Json { ignoreUnknownKeys = true; isLenient = true }

    // -- Quote normalisation -----------------------------------------------

    @Test
    fun `normalizeQuote turns raw quote fixture into normalised fixture`() {
        val rawJson = loadFixture("quote.json")
        val expectedJson = loadFixture("quote.normalized.json")

        val raw = jsonParser.decodeFromString<RawQuote>(rawJson)
        val expected = jsonParser.parseToJsonElement(expectedJson) as kotlinx.serialization.json.JsonObject

        val q = normalizeQuote(raw)

        assertEquals(expected["symbol"]!!.asPrimitive().content, q.symbol)
        assertEquals(expected["name"]!!.asPrimitive().content, q.name)
        assertEquals(expected["price"]!!.asPrimitive().double, q.price)
        assertEquals(expected["change"]!!.asPrimitive().double, q.change)
        assertEquals(expected["open"]!!.asPrimitive().double, q.open)
        assertEquals(expected["high"]!!.asPrimitive().double, q.high)
        assertEquals(expected["low"]!!.asPrimitive().double, q.low)
        assertEquals(expected["volume"]!!.asPrimitive().long, q.volume)
        assertEquals(expected["marketCap"]!!.asPrimitive().double, q.marketCap)
        assertEquals(expected["currency"]!!.asPrimitive().content, q.currency)
        assertEquals(expected["exchange"]!!.asPrimitive().content, q.exchange)

        // changePercent: floating-point tolerance
        assertEquals(
            expected["changePercent"]!!.asPrimitive().double,
            q.changePercent,
            1e-6,
            "changePercent should match within tolerance"
        )

        // timestamp: ISO string matches fixture
        assertEquals(
            expected["timestamp"]!!.asPrimitive().content,
            q.toMap()["timestamp"],
        )
    }

    @Test
    fun `normalizeQuote returns zero percent when close is zero`() {
        val rawJson = loadFixture("quote.json")
        val raw = jsonParser.decodeFromString<RawQuote>(rawJson)
        val modified = raw.copy(
            change = 10.0,
            ohlc = raw.ohlc.copy(close = 0.0),
        )

        val q = normalizeQuote(modified)

        assertEquals(0.0, q.changePercent)
    }

    @Test
    fun `normalizeQuote converts non-UTC offset to UTC`() {
        val rawJson = loadFixture("quote.json")
        val raw = jsonParser.decodeFromString<RawQuote>(rawJson)
        // 15:45 IST (+05:30) is exactly the 10:15 UTC the fixture already encodes.
        val modified = raw.copy(timestamp = "2026-05-27T15:45:00.000+05:30")

        val q = normalizeQuote(modified)

        assertEquals("2026-05-27T10:15:00.000Z", q.toMap()["timestamp"])
    }

    // -- Search / instrument normalisation ---------------------------------

    @Test
    fun `normalizeInstrument turns raw search fixture into normalised results`() {
        val rawJson = loadFixture("search.json")
        val expectedJson = loadFixture("search.normalized.json")

        val raw = jsonParser.decodeFromString<RawSearchResponse>(rawJson)
        val expected = jsonParser.parseToJsonElement(expectedJson) as kotlinx.serialization.json.JsonObject
        val expectedResults = expected["results"] as kotlinx.serialization.json.JsonArray

        val result = raw.instruments.map { normalizeInstrument(it) }

        assertEquals(expectedResults.size, result.size)
        result.forEachIndexed { i, instrument ->
            val exp = expectedResults[i] as kotlinx.serialization.json.JsonObject
            assertEquals(exp["symbol"]!!.asPrimitive().content, instrument.symbol)
            assertEquals(exp["name"]!!.asPrimitive().content, instrument.name)
            assertEquals(exp["exchange"]!!.asPrimitive().content, instrument.exchange)
            assertEquals(exp["type"]!!.asPrimitive().content, instrument.type.value)
            assertEquals(exp["currency"]!!.asPrimitive().content, instrument.currency)
        }
    }
}

// Extension to make JSON assertions cleaner
private fun kotlinx.serialization.json.JsonElement.asPrimitive() =
    this as kotlinx.serialization.json.JsonPrimitive
